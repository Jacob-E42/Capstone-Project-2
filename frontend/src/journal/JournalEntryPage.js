import React, { useCallback, useContext, useEffect, useRef } from "react";
import { useNavigate, useParams, Navigate } from "react-router-dom";
import Journal from "./Journal";
import LoadingSpinner from "../common/LoadingSpinner";
import UserContext from "../context_providers/UserContext";
import AlertContext from "../context_providers/AlertContext";
import ApiContext from "../context_providers/ApiContext";
import useLocalStorage from "../hooks/useLocalStorage";
import { getCurrentDate } from "../common/dateHelpers";
import StreakDisplay from "../streak/StreakDisplay";
import Feedback from "../feedback/Feedback";
import Emotions from "../emotions/Emotions";

const verifyDependentInfo = (date, user, api) => {
	if (!(date && user && api)) return false;
	if (typeof date !== "string") return false;
	if (date.length < 10) return false;
	if (typeof user !== "object") return false;
	if (typeof api !== "object") return false;
	return true;
};

const validateJournalInfo = (id, userId, title, entryText, date, journalType) => {
	console.debug("validateJournalInfo", id, userId, title, entryText, date, journalType);
	if (!id || !userId || !title || !entryText || !date || !journalType)
		return { valid: false, error: "REQUIRED INFO IS MISSING" };
	if (typeof id !== "number") return { valid: false, error: "ID IS NOT A NUMBER" };
	if (typeof userId !== "number") return { valid: false, error: "USERID IS NOT A NUMBER" };
	if (typeof title !== "string") return { valid: false, error: "TITLE IS NOT A STRING" };
	if (typeof date !== "string") return { valid: false, error: "DATE IS NOT A STRING" };
	if (typeof entryText !== "string") return { valid: false, error: "ENTRYTEXT IS NOT A STRING" };
	if (typeof journalType !== "string") return { valid: false, error: "JOURNALTYPE IS NOT A STRING" };
	if (entryText.trim().length < 1) return { valid: false, error: "ENTRYTEXT IS TOO SHORT" };
	if (date.length !== 10) return { valid: false, error: "DATE IS WRONG LENGTH" };
	if (!journalType.includes("Journal")) return { valid: false, error: "JOURNAL TYPE IS WRONG" };

	return { valid: true };
};

const JournalEntryPage = () => {
	let { date } = useParams("date");
	if (!date) date = getCurrentDate();
	const { user } = useContext(UserContext);
	const { api } = useContext(ApiContext);
	const { setMsg, setColor } = useContext(AlertContext);
	const allInfoDefined = verifyDependentInfo(date, user, api); //only verifies date, user, and qpi. Not setMsg, or setColor
	const navigate = useNavigate();
	const [currentJournal, setCurrentJournal] = useLocalStorage("currentJournal", null);
	const [journalLoaded, setJournalLoaded] = useLocalStorage("journalLoaded", false);
	const [feedbackPending, setFeedbackPending] = useLocalStorage("feedbackPending", false);
	const [feedback, setFeedback] = useLocalStorage("feedback", null);
	const [feedbackReceived, setFeedbackReceived] = useLocalStorage("feedbackReceived", false);
	const [emotionsReceived, setEmotionsReceived] = useLocalStorage("emotionsReceived", false);
	const [emotionsPending, setEmotionsPending] = useLocalStorage("setEmotionsPending", false);
	const lastVisitedPage = useRef(getCurrentDate());

	console.debug(
		"JournalEntryPage",
		"date=",
		date,
		"user=",
		user,
		"api=",
		api,
		"allInfoDefined=",
		allInfoDefined,
		"journal=",
		currentJournal,
		"journalLoaded=",
		journalLoaded,
		"feedbackPending=",
		feedbackPending,
		"feedbackReceived=",
		feedbackReceived,
		"emotionsReceived=",
		emotionsReceived,
		"emotions=",
		currentJournal?.emotions
	);

	useEffect(() => {
		console.debug("useEffect - JournalEntryPage", "date=", date, "currentJournal=", currentJournal);

		if (!allInfoDefined) {
			setJournalLoaded(false);
		} else {
			setFeedback(null);
			setCurrentJournal(null);
			loadJournalEntry();
		}
		// eslint-disable-next-line
	}, [date, api]);

	// ?? This causes a rerender whenever currentJournal is updated. Currently only when loadJournalEntry or
	// when editJournal are called.
	// useEffect(() => {}, [currentJournal]);

	const loadJournalEntry = useCallback(async () => {
		console.debug("loadJournalEntry");
		try {
			// console.log("date=", date, "api=", api);
			const resp = await api.getJournalEntryByDate(user.id, date);
			console.debug("Here is the RESPONSE", resp);
			if (typeof resp !== "object" || resp.status === 404) throw new Error("Response returned was invalid");
			setCurrentJournal(resp);
			await setJournalLoaded(true);
			lastVisitedPage.current = date;
		} catch (err) {
			console.error(err, err.status);
			setMsg(err.message);
			setColor("error");
			if (err.status === 404) {
				setJournalLoaded(true);
			} else {
				setJournalLoaded(false);
			}
			setCurrentJournal(null);
		}
	}, [setMsg, setColor, api, date, user, setCurrentJournal, setJournalLoaded, navigate]);

	const editJournal = useCallback(
		async data => {
			console.debug("JournalEntryPage editJournal", "currentJournal=", currentJournal, "data=", data);
			setJournalLoaded(false);
			if (!currentJournal) {
				setMsg("Creating a new journal entry failed!");
				setColor("error");
			} else {
				await setFeedbackPending(true);
				await setEmotionsPending(true);
				try {
					// console.log(currentJournal, data);

					const updateJournal = await api.editJournalEntry(
						currentJournal.userId,
						data.title,
						data.entryText,
						currentJournal.entryDate,
						currentJournal.journalType
					);
					// console.log(updateJournal);
					if (updateJournal) {
						setMsg("Journal entry updated!");
						setColor("success");
						setJournalLoaded(true);
					}
				} catch (err) {
					// console.log(err);
					setJournalLoaded(true);
				}
			}
		},
		[currentJournal, setColor, setMsg, api, setJournalLoaded, setFeedbackPending, setEmotionsPending]
	);

	const fetchFeedback = useCallback(() => {
		console.debug("fetchFeedback");
		async function loadFeedback() {
			setFeedbackReceived(false);
			const { id, userId, title, entryText, journalType } = currentJournal;
			const validJournal = validateJournalInfo(id, userId, title, entryText, date, journalType);
			if (validJournal.valid) {
				try {
					console.debug("Journal is valid");
					const feedback = await api.getFeedback(id, userId, entryText, journalType, title, date);
					if (feedback && !feedback.error) {
						setMsg("Feedback Received!");
						setColor("success");
						setFeedback(feedback);
						setFeedbackReceived(true);
					} else throw feedback.error;
				} catch (err) {
					console.error(err);
					setMsg("Loading Feedback Failed");
					setColor("error");
				}
			} else {
				setMsg("Journal is NOT valid");
				console.debug(validJournal.error);
			}
		}
		loadFeedback();

		setFeedbackPending(false);
	}, [api, currentJournal, setMsg, setColor, date, setFeedback, setFeedbackReceived, setFeedbackPending]);

	useEffect(() => {
		console.debug("useEffect -> fetchFeedback()");
		if (currentJournal && currentJournal.entryText && feedbackPending) {
			fetchFeedback();
		} else {
			console.warn("FEEDBACK IS NOT PENDING", feedbackPending, currentJournal?.entryText);
			setMsg("An error occurred trying to load feedback.");
			setColor("error");
		}
		// eslint-disable-next-line
	}, [feedbackPending]);

	const fetchEmotions = useCallback(() => {
		console.debug("fetchEmotions");
		async function loadEmotions() {
			setEmotionsReceived(false);
			const { id, userId, title, entryText, journalType } = currentJournal;
			const validJournal = validateJournalInfo(id, userId, title, entryText, date, journalType);
			if (validJournal.valid) {
				try {
					console.debug("Journal is valid");
					const resp = await api.getEmotions(id, userId, entryText, journalType, title, date);
					console.log(resp);

					if (resp) {
						setMsg("Emotions Received!");
						setColor("success");
						const newJournal = currentJournal;
						newJournal["emotions"] = resp;
						setCurrentJournal(newJournal);
						setEmotionsReceived(true);
					}
				} catch (err) {
					console.error(err);
					setMsg("Loading Emotions Failed");
					setColor("error");
				}
			} else {
				setMsg("Journal is NOT valid");
				console.debug(validJournal.error);
			}
		}
		loadEmotions();
		setEmotionsPending(false);
	}, [api, currentJournal, setMsg, setColor, date, setCurrentJournal, setEmotionsReceived, setEmotionsPending]);

	useEffect(() => {
		console.debug("useEffect -> fetch Emotions()");
		if (currentJournal && currentJournal.entryText && emotionsPending) {
			fetchEmotions();
		} else {
			console.warn("Emotions Are NOT PENDING", emotionsPending, currentJournal?.entryText);
		}
		// eslint-disable-next-line
	}, [emotionsPending]);

	if (!journalLoaded) return <LoadingSpinner />;

	console.log("test allInfoDefined & currentJournal", allInfoDefined, currentJournal, journalLoaded);

	return (
		<>
			<StreakDisplay
				date={date}
				userId={user.id}
				api={api}
			/>
			{allInfoDefined && currentJournal && (
				<Journal
					date={date}
					title={currentJournal.title}
					entryText={currentJournal.entryText}
					jounalType={currentJournal.journalType}
					setJournal={setCurrentJournal}
					currentJournal={currentJournal}
					editJournal={editJournal}
				/>
			)}
			{allInfoDefined && !currentJournal && date && (
				<Navigate
					to={`/journal/${lastVisitedPage.current}`}
					replace
				/>
			)}

			{feedbackReceived && currentJournal && <Feedback feedback={feedback} />}
			{emotionsReceived && currentJournal && <Emotions emotions={currentJournal?.emotions} />}
		</>
	);
};

export default JournalEntryPage;

// Development Practices
// The way you use the library can have a bigger impact on performance than the library itself. For instance,
// lazy loading components, optimizing re-renders,
//  and properly managing state can lead to significant performance improvements regardless of the UI library you choose.
