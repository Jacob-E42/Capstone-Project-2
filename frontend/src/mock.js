import React, { useCallback, useState } from "react";
import UserContext from "./context_providers/UserContext";
import AlertContext from "./context_providers/AlertContext";
import Request from "./api";
import ApiRequest from "./api";
import ApiContext from "./context_providers/ApiContext";
import JournalContext from "./context_providers/JournalContext";

const demoUser = {
	id: 11,
	firstName: "testfirst",
	lastName: "testlast",
	email: "test@test.net",
	interests: ["baking", "jetskiing"]
};

const demoJournal = {
	id: 42,
	userId: 11,
	title: "Baking adventure",
	entryText:
		"Yesterday I tried baking for the first time like I've always wanted to. It was a complete disaster, but at least it was fun.",
	entryDate: "2023-07-24",
	emotions: null
};

const UserProvider = ({ children }) => {
	return <UserContext.Provider value={{ user: demoUser }}>{children}</UserContext.Provider>;
};

const AnonUserProvider = ({ children }) => {
	const [currentUser, setCurrentUser] = useState(null);

	const login = useCallback(
		async formData => {
			try {
				const request = new Request();
				const token = await request.login(formData);
				if (token) {
					setCurrentUser(demoUser);
					return { success: true };
				} else throw new Error("Log in token in missing");
			} catch (errors) {
				console.error("login failed", errors);
				return { success: false, errors };
			}
		},
		[setCurrentUser]
	);

	const signup = useCallback(
		async formData => {
			try {
				const request = new Request();
				const token = await request.signup(formData);
				console.log(token);
				if (token) setCurrentUser(demoUser);
				else throw new Error("missing sign in token");
				return { success: true };
			} catch (errors) {
				console.error("signup failed", errors);
				return { success: false, errors };
			}
		},
		[setCurrentUser]
	);

	return <UserContext.Provider value={{ currentUser, login, signup }}>{children}</UserContext.Provider>;
};

const AlertProvider = ({ children }) => {
	const [msg, setMsg] = useState("");
	const [color, setColor] = useState("primary");

	return <AlertContext.Provider value={{ msg, setMsg, color, setColor }}>{children}</AlertContext.Provider>;
};

const ApiProvider = ({ children }) => {
	const [request, setRequest] = useState(new ApiRequest());
	request.setToken(
		`eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MTEsImVtYWlsIjoidGVzdEB0ZXN0Lm5ldCIsImlhdCI6MTY5MjI5MDkxN30.8e12Cw-TMts0FmZALH61ch03kUMlW5QoCi60dlfhyWY`
	);

	return <ApiContext.Provider value={{ api: request }}>{children}</ApiContext.Provider>;
};

const JournalProvider = ({ children }) => {
	return <JournalContext.Provider value={{ journal: demoJournal }}>{children}</JournalContext.Provider>;
};

export { UserProvider, AnonUserProvider, AlertProvider, ApiProvider, JournalProvider };
