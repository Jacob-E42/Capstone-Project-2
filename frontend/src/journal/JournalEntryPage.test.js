import React from "react";
import { render, screen, act } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { AlertProvider, AnonUserProvider, ApiProvider, JournalProvider, UserProvider } from "../mock";
import JournalEntryPage from "./JournalEntryPage";
import axios from "axios";
jest.mock("axios");

const mockSuccessfulResponse = {
	data: {
		journal: {
			id: 42,
			userId: 11,
			title: "Baking adventure",
			entryText:
				"Yesterday I tried baking for the first time like I've always wanted to. It was a complete disaster, but at least it was fun.",
			entryDate: "2023-07-24",
			emotions: null
		}
	}
};

const mockErrorResponse = {
	response: {
		status: 404,
		data: {
			error: {
				message: `No journal with date: 2020-07-24`
			}
		}
	}
};

describe("Successful journal fetch", () => {
	beforeEach(() => {
		jest.clearAllMocks();
		act(() => {
			axios.mockResolvedValue(mockSuccessfulResponse);
		});
	});

	test("JournalEntryPage renders without crashing", () => {
		render(
			<MemoryRouter>
				<UserProvider>
					<JournalProvider>
						<ApiProvider>
							<AlertProvider>
								<JournalEntryPage />
							</AlertProvider>
						</ApiProvider>
					</JournalProvider>
				</UserProvider>
			</MemoryRouter>
		);
	});

	test("JournalEntryPage rendered correctly matches snapshot", () => {
		const { asFragment } = render(
			<MemoryRouter>
				<UserProvider>
					<JournalProvider>
						<ApiProvider>
							<AlertProvider>
								<JournalEntryPage />
							</AlertProvider>
						</ApiProvider>
					</JournalProvider>
				</UserProvider>
			</MemoryRouter>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	test("JournalEntryPage renders expected text", async () => {
		// eslint-disable-next-line testing-library/no-unnecessary-act
		await act(async () => {
			render(
				<MemoryRouter>
					<UserProvider>
						<JournalProvider>
							<ApiProvider>
								<AlertProvider>
									<JournalEntryPage propDate="2023-07-04" />
								</AlertProvider>
							</ApiProvider>
						</JournalProvider>
					</UserProvider>
				</MemoryRouter>
			);
		});

		expect(screen.getByText("2023-07-04")).toBeInTheDocument();
		expect(screen.getByText("Journal Entry")).toBeInTheDocument();
		expect(screen.getByPlaceholderText("title")).toBeInTheDocument();
		expect(screen.getByPlaceholderText("Start your entry here...")).toBeInTheDocument();
	});
});

describe("Failed journal fetch", () => {
	beforeEach(() => {
		jest.clearAllMocks();
		act(() => {
			axios.mockRejectedValue(mockErrorResponse);
		});
	});

	test("JournalEntryPage crashed, matches snapshot", () => {
		const { asFragment } = render(
			<MemoryRouter>
				<UserProvider>
					<JournalProvider>
						<ApiProvider>
							<AlertProvider>
								<JournalEntryPage propDate={42} />
							</AlertProvider>
						</ApiProvider>
					</JournalProvider>
				</UserProvider>
			</MemoryRouter>
		);

		const allMatches = screen.queryAllByText("Error: A date must be provided");
		expect(allMatches.length).toBeGreaterThan(0);
		expect(asFragment()).toMatchSnapshot();
	});

	test("JournalEntryPage renders expected text when the date passed doesn't have an associated journal entry", async () => {
		// eslint-disable-next-line testing-library/no-unnecessary-act
		await act(async () => {
			render(
				<MemoryRouter>
					<UserProvider>
						<JournalProvider>
							<ApiProvider>
								<AlertProvider>
									<JournalEntryPage propDate="2020-07-04" />
								</AlertProvider>
							</ApiProvider>
						</JournalProvider>
					</UserProvider>
				</MemoryRouter>
			);
		});

		expect(screen.getByText("Error: There is no journal entry for date: 2020-07-04")).toBeInTheDocument();
	});

	test("JournalEntryPage renders expected text when allInfoDefined is false", async () => {
		// eslint-disable-next-line testing-library/no-unnecessary-act

		const { asFragment } = render(
			<MemoryRouter>
				<AnonUserProvider>
					<JournalProvider>
						<ApiProvider>
							<AlertProvider>
								<JournalEntryPage propDate="2023-07-04" />
							</AlertProvider>
						</ApiProvider>
					</JournalProvider>
				</AnonUserProvider>
			</MemoryRouter>
		);

		expect(asFragment()).toMatchSnapshot();
		const allMatches = screen.queryAllByText("Error: A date must be provided");
		expect(allMatches.length).toBeGreaterThan(0);
	});
});
