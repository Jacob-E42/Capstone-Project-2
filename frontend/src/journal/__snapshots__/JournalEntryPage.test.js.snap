// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JournalEntryPage crashed, matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="alert alert-danger fade"
    role="alert"
  >
    There was an error loading the journal entry
  </div>
  <p>
    Streak goes here
  </p>
  <p>
    feedback goes here
  </p>
</DocumentFragment>
`;

exports[`JournalEntryPage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="alert alert-danger fade"
    role="alert"
  >
    A journal could not be found
  </div>
  <p>
    Streak goes here
  </p>
  <p>
    feedback goes here
  </p>
</DocumentFragment>
`;

exports[`JournalEntryPage rendered correctly matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="alert alert-danger fade"
    role="alert"
  >
    There was an error loading the journal entry
  </div>
  <p>
    Streak goes here
  </p>
  <p>
    feedback goes here
  </p>
</DocumentFragment>
`;
